Add support for Asciidoc in configuration reference documentation

gsmet opened the pull request:
/cc @emmanuelbernard @FroMage @machi1990 

Here it is. It's a bit more than a one liner as I had to restructure the parser and the output a bit. It fixes a logic issue anyway so it's all for good.

I haven't written the extension guide documentation yet but let's agree on this first.

One thing that is a bit less nice compared to the handcrafted output is that we lost the nice titled sections but I think having everything generated is worth it.

At some point, we could imagine a new tag to generate proper sections but let's do that another day.

Once Asciidoclet has the related patch, we will have a proper output in the generated Javadoc too. 

@machi1990 could you upgrade your repository with this PR and point to it so that Emmanuel can have a look?

Note that GitHub does not support the new Font Awesome icons, I will attach a few screenshots.

### Summary table

![picture1](https://user-images.githubusercontent.com/1279749/64363723-a1dac000-d011-11e9-8fdd-99241e9978e3.png)

### Details with Asciidoc

![picture2](https://user-images.githubusercontent.com/1279749/64363727-a606dd80-d011-11e9-83d8-3505555fb62b.png)

gsmet committed:
Restructure the parser to allow passing optional Asciidoc comment

Asciidoc comment are passthrough but we needed to restructure the parser
to enable proper parsing.

gsmet committed:
Update the Hibernate ORM config javadoc with the content of the guide

machi1990 commented:
> @machi1990 could you upgrade your repository with this PR and point to it so that Emmanuel can have a look?

Repository update - https://github.com/machi1990/quarkus-generated-extension-config-docs/blob/master/src/main/asciidoc/generated/quarkus-hibernate-orm.adoc 

> Note that GitHub does not support the new Font Awesome icons, I will attach a few screenshots. 

Good thing you attached screenshots. GH did not like the icons :-P

machi1990 commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfig.java:
```diff
@@ -117,27 +151,29 @@ public boolean isAnyPropertySet() {
+         *
+         * Accepted values: `none`, `create`, `drop-and-create`, `drop`, `update`. The default is `none`.
          */
         @ConfigItem
```
Maybe add a `defaultValue="none"` to the annotation so that it is nicely added to the summary table :-)

gsmet replied:
Yes, I have to tweak things a bit regarding that. And wondering if we could use the enum directly now but I have to test it. It's on my list of things to improve.

machi1990 replied:
An enum would be even better.

gsmet replied:
Yup, I'm just not sure it will fly, I need to check it and wanted this PR out of the door.

I will fix the default value though. Small improvements are better than later improvements :).

gsmet replied:
Done for the default value.

gsmet committed:
Use defaultValue to define the default value of ORM schema generation

This way, it's properly documented.

FroMage commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfig.java:
```diff
@@ -95,15 +127,17 @@ public boolean isAnyPropertySet() {
-         * Valid values are: none, first, last.
+         * Default precedence of null values in `ORDER BY` clauses.
+         *
+         * Valid values are: `none`, `first`, `last`.
```
Shouldn't we replace this `String` type with an enum?

gsmet replied:
Sure, see the discussion with Manyanda above. But I need to check if we can use an ORM one or if we have to create a new one so I consider it outside of the scope of this PR whose main purpose is to validate the Asciidoc support.

FroMage replied:
Ah sorry I didn't see it. Sure it can come in another PR.

FroMage commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfig.java:
```diff
@@ -117,27 +151,29 @@ public boolean isAnyPropertySet() {
+         *
+         * `drop-and-create` is awesome in development mode.
+         *
+         * Accepted values: `none`, `create`, `drop-and-create`, `drop`, `update`. The default is `none`.
```
Same here, this custom tag makes it impossible to extract the allowed values in the table, so probably should use an enum, no?

gsmet committed:
Unify EOL on all platforms when returning the original Asciidoc

The Javadoc parser we use uses platform specific EOL characters.

This should fix the Windows CI failure.

gsmet committed:
Add a light separator after each configuration entry in the details

Note that we will have to tweak the website CSS to keep it light.

machi1990 commented:
> I improved the output a bit by adding a light separator between each entry of the detailed output. Things are far more readable this way.

tracking progress repo updated

gsmet committed:
Remove leading space in the extracted Javadoc

The space just after the star is kept by the javax.lang.model API so
let's remove it.

gsmet committed:
Prepend lines with a star as it's what is expected by the parser

gsmet committed:
Tweak Hibernate ORM config following the latest adjustments

gsmet committed:
Move the generated documentation to a global target/ directory

We can't have the generated files in docs/target as it is cleaned prior
to generating the documentation when doing a mvn clean install.

Fixes #3909

emmanuelbernard commented:
Boom, its in!
